doc/developer/design/20230418_persist_state_pubsub.md

```diff
@@ -0,0 +1,348 @@
+for discovering state updates, though with a more generous retry policy than we currently have. If a process receives a
+diff that does not apply cleanly (e.g. out-of-order delivery), the process will fetch the latest state from Cockroach
+directly instead. While we will continue to have these fallbacks, we anticipate message delivery being effective enough
+to eliminate the need for nearly all Cockroach reads.
+
+# Reference Explanation
+
+As a first implementation of the Persist sidechannel, we propose building a new RPC service that flows through `environmentd`.
+The service would be backed by gRPC with a minimal surface area -- to start, we would use a simple bidirection stream that
+allows each end to push updates back and forth, with `environmentd` broadcasting out any changes it receives.
+
+## RPC Service
+
+```protobuf
+message ProtoPushDiff {
+  string shard_id = 1;
+  uint64 seqno = 2;
+  bytes diff = 3;
+}
+
+message ProtoSubscribe {
+  string shard = 1;
+}
+
+message ProtoUnsubscribe {
+  string shard = 1;
+}
+
+message ProtoPubSubMessage {
+  oneof message {
+    ProtoPushDiff push_diff = 1;
+    ProtoSubscribe subscribe = 2;
+    ProtoUnsubscribe unsubscribe = 3;
+  }
+}
+
+service ProtoPersistPubSub {
+  rpc PubSub (stream ProtoPubSubMessage) returns (stream ProtoPubSubMessage);
+}
+```
+
+The proposed topology would have each `clusterd` connect to `environmentd`'s RPC service. `clusterd` would subscribe to
+updates to any shards in its `PersistClientCache` (a singleton in both `environmentd` and `clusterd`). When any `clusterd`
+commits a change to a shard, it publishes the diff to `environmentd`, which tracks which connections are subscribed to which
+shards, and broadcasts out the diff to the interested subscribers. From there, each subscriber would [apply the diff](#applying-received-updates).
+Note that a `clusterd` would only subscribe to a shard if it is actively reading it, e.g. for a materialized view, `SELECT`,
+or a `SUBSCRIBE`, so the number of broadcasted messages would be proportional to the number of ongoing read operations,
+and not the total number of shards.
+
+`environmentd` is chosen to host the RPC service as it necessarily holds a handle to each shard (and therefore is always
+interested in all diffs), and to align with the existing topology of the Controller, rather than introduce the novel
+complexity of point-to-point connections between unrelated `clusterd` processes or a net-new dedicated process. With this
+implementation we would anticipate the end-to-end latency between a writer in one `clusterd` pushing an update to a reader
+in another `clusterd` receiving that update to be in the 1-10ms range.
+
+### Service Discovery
+
+`clusterd` will need to know the address of the `environmentd` process in order to connect to the PubSub server. Locally,
+this will be done through a standard `127.0.0.1:6879` port and passed in to `clusterd` via command line argument.
```

Comment:
I don't understand this sentence. What do you mean by "Locally", and what is done on `127.0.0.1:6879`?

Reply:
ah, I meant when doing local development we can have `envd` bind the server to `127.0.0.1:6879` and point `clusterd` to that address

Reply:
Maybe we could rephrase this to make it clear that this is for local development? The term "locally" is unclear.

Reply:
s/Locally/When using the process orchestrator/?